### PR TITLE
Add test case test_migration_with_rebuilding_replica

### PR DESCRIPTION
**Pull request by chriscchien**
Signed-off-by: Chris Chien <chris.chien@suse.com>

Add test case test_migration_with_rebuilding_replica

### CODE REVIEW

File: manager/integration/tests/test_migration.py [hunk 1]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
...
+                                  numberOfReplicas=3,
+                                  backingImage="",
+                                  accessMode="rwx", migratable=True)
+    volume = common.wait_for_volume_detached(client, volume_name)
+    volume.attach(hostId=current_host)
+    volume = common.wait_for_volume_healthy(client, volume_name)
+
+    volume_endpoint = get_volume_endpoint(volume)
+    write_volume_dev_random_mb_data(volume_endpoint,
+                                    1, 1500)
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_migration.py:242]: Is 1M data enough for replica to rebuild while the volume got attached to another node and we can assert 'there should be no replica created before rebuilding complete' ?

File: manager/integration/tests/test_migration.py [hunk 2]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
...
+            if r.name != replica0.name and \
+                    r.name != replica1.name and \
+                    r.name != replica2.name:
+                new_replica_found = True
+                break
+        if new_replica_found:
+            break
+        time.sleep(RETRY_INTERVAL)
+
+    assert new_replica_found is True
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_migration.py:265]: In the step 3, a replica has to be removed only. the new replicas for migration will get rebuild only after this volume is attached to another node.

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_migration.py (hunk 1 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py:242]: Hello @khushboo-rancher , 
I use 1500M to make sure the volume attached on origin node is in rebuilding status and then we can do migration to target node. If use small data size, rebuild on origin node will finish soon and hard to match the scenario test case wanted.

File: manager/integration/tests/test_migration.py (hunk 2 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py:265]: Hi @khushboo-rancher ,

Above code is to make sure replica rebuilding happened on volume which not migrate yet, if I missed something, I will update soon, thank you

**Comment by innobead** [comment]
@chriscchien Please resolve the conflicts from the branch first.

File: manager/integration/tests/test_migration.py [hunk 3]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
     7. Observe volume migrated to node 2 (single active engine).
     8. Validate initially written test data.
     """
-    return
+    # Step 1
+    client = get_random_client(clients)  # NOQA
```
Comment (innobead) [inline manager/integration/tests/test_migration.py:227]: Why we need to add `# NOQA`, also this commit looks weird and not sure the purpose `sdf`?.

File: manager/integration/tests/test_migration.py [hunk 4]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
     7. Observe volume migrated to node 2 (single active engine).
     8. Validate initially written test data.
     """
-    return
+    # Step 1
+    client = get_random_client(clients)
+    current_host = common.get_self_host_id()
+    host1, host2 = get_hosts_for_migration_test(clients) # NOQA
```
Comment (innobead) [inline manager/integration/tests/test_migration.py:229]: Why keep adding `# NOQA`? any purpose?

File: manager/integration/tests/test_migration.py [hunk 5]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
...
+                new_replica_found = True
+                break
+        if new_replica_found:
+            break
+        time.sleep(RETRY_INTERVAL)
+
+    assert new_replica_found is True
+
+    # Step 4
+    volume.attach(hostId=host1)
```
Comment (innobead) [inline manager/integration/tests/test_migration.py:268]: As per the step 4, need to see new replicas created after the removed replica gets rebuilt successfully first.

File: manager/integration/tests/test_migration.py [hunk 6]
```diff
@@ -220,7 +223,70 @@ def test_migration_with_rebuilding_replica(clients, volume_name):  # NOQA
...
+
+    # Step 5
+    volume = common.wait_for_volume_migration_ready(client, volume_name)
```
Comment (innobead) [inline manager/integration/tests/test_migration.py:271]: Need to check the number of replicas from the first volume the same as the second migrated.

**Review by innobead — CHANGES_REQUESTED** [review]
Please ensure the implementation based on the test steps.

File: manager/integration/tests/test_migration.py (hunk 2 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py]: I update this part of code by using `wait_for_rebuild_start` to reuse exist function

File: manager/integration/tests/test_migration.py (hunk 4 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py]: Sorry, I removed the unnecessary comment

File: manager/integration/tests/test_migration.py (hunk 5 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py:321]: Hi @innobead , I added `wait_for_rebuild_start` and `wait_for_rebuild_complete` to check the delete replica rebuild process

File: manager/integration/tests/test_migration.py (hunk 6 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py:326]: I update code to check the new replica numbers for review, thank you.

File: manager/integration/tests/test_migration.py (hunk 5 above)
Comment (khushboo-rancher) [reply manager/integration/tests/test_migration.py:321]: @chriscchien I'm not sure, I understood. I think we need to check the replica count as `old_replicas` until the rebuilding is completed.

**Review by khushboo-rancher — COMMENTED** [review]
LGTM in general, one last comment.

File: manager/integration/tests/test_migration.py (hunk 5 above)
Comment (chriscchien) [reply manager/integration/tests/test_migration.py:321]: Hi @khushboo-rancher , you are right, I add check code at line 323 to check replica count. but it's not to easy checkout the count during rebuild dynamically, so I checked once, thank you

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@chriscchien Please backport to 1.3 and 1.2. Thanks.

**Pull request merged**